### ISSUE TITLE
Support non-US keyboard layouts

### ISSUE THREAD

**Issue by jasongrout** [jupyterlab/jupyterlab#7579]
There are a number of issues with using JupyterLab with non-US keyboards. To summarize, we have two keyboard shortcut systems:

* Lumino (phosphor) keyboard shortcut system, used throughout JupyterLab. This shortcut system allows for non-US keyboards, but we need to have relevant keyboard mappings. We've talked about writing some UX to make it easier to make a keymap (basically capturing the key character and usb code and associating them, perhaps combining maps from different browsers)
* codemirror keyboard shortcut system for text editors and code cells. Should we expose this in some way that integrates with the Lumino shortcut system? Or should we expose this as a separate mechanism?

Designing a UX for capturing new keyboard mappings would be a great new extension, for example.

## Related issues

#5164

**Comment by vidartf** [comment]
Would you mind expanding a little on why a keymap is needed / what problems they would solve?

**Comment by calvin-sykes** [comment]
Hi - just to point out a related issue is here: https://github.com/jupyterlab/jupyterlab/issues/4778

Forgive the unsolicited comment, but it does seem somewhat bizarre that a web app need concern itself with the mapping between physical keys and characters. Naively I would've assumed that this would remain the OS' and/or browser's job.

**Comment by cstenkamp** [comment]
> Would you mind expanding a little on why a keymap is needed / what problems they would solve?

Many Non-US-Keyboards don't have certain buttons, like the "/". So on these keyboards, it is impossible to for example use the "comment code block"-functionality, which is really annoying.

**Comment by jasongrout** [comment]
> Naively I would've assumed that this would remain the OS' and/or browser's job.

How we wish that were true! When we were designing the Lumino keyboard system a few years ago, the browsers were woefully inconsistent about how they applied these settings, and indeed still are last I checked, so we had to go with a lower common denominator of keycodes rather than relying on the browser giving us accurate key character information. Unfortunately, this means that we need to explicitly give the mapping between keycodes and actual silkscreened characters.

**Comment by jasongrout** [comment]
> Many Non-US-Keyboards don't have certain buttons, like the "/". So on these keyboards, it is impossible to for example use the "comment code block"-functionality, which is really annoying.

FYI, those specific keys are set here:

https://github.com/jupyterlab/jupyterlab/blob/4fe4dcfe5c9dc329bed2dcf2602f569ddef8a8a0/packages/codemirror/src/factory.ts#L24-L25

https://github.com/jupyterlab/jupyterlab/blob/4fe4dcfe5c9dc329bed2dcf2602f569ddef8a8a0/packages/codemirror/src/factory.ts#L34-L35

I think it makes sense to have those specific bindings configurable via settings or JLab's shortcut system, if someone wants to submit a PR.

**Comment by vidartf** [comment]
> > Would you mind expanding a little on why a keymap is needed / what problems they would solve?
>
> Many Non-US-Keyboards don't have certain buttons, like the "/". So on these keyboards, it is impossible to for example use the "comment code block"-functionality, which is really annoying.

I agree that this is a problem, but I'm saying that I'm not sure if I see how a keymap will solve it. Are we planning to special case the most popular maps, or is there some general way to solve the problem given a keymap?

**Comment by vidartf** [comment]
For context, I use a Norwegian keyboard layout, but use Window's Alt+Shift shortcut to toggle to a US layout for coding so that the `øæå¨'` keys next to Enter map to `;'[]\` instead of having to use Alt Gr + 8 / 9 for square brackets. I'm not sure if I want Lab to interfere in this process.

**Comment by jasongrout** [comment]
The current plan in Lumino is that it will maintain a database of keymaps and the user will be able to select the appropriate one to use manually. Perhaps there could be some autoguessing logic like is popular in various OS installations (like press the key to the right of Shift, etc.).

Here is the current keymap Lumino has: https://github.com/jupyterlab/lumino/blob/master/packages/keyboard/src/index.ts#L210

**Comment by jasongrout** [comment]
> When we were designing the Lumino keyboard system a few years ago, the browsers were woefully inconsistent about how they applied these settings, and indeed still are last I checked, so we had to go with a lower common denominator of keycodes rather than relying on the browser giving us accurate key character information. Unfortunately, this means that we need to explicitly give the mapping between keycodes and actual silkscreened characters.

I just wrote up more explicit details about this in https://github.com/jupyterlab/lumino/issues/74

**Comment by jweill-aws** [comment]
@jasongrout noted that Chrome can now provide info about the user's keyboard layout to web applications. Other browsers do not yet support this. To cover all users, provide an option for the user to manually set their keyboard layout.

**Comment by jasongrout** [comment]
> @jasongrout noted that Chrome can now provide info about the user's keyboard layout to web applications.

See https://github.com/jupyterlab/lumino/issues/271

**Comment by krassowski** [comment]
A minimal improvement that we could make would be ensuring that if a default shortcut conflicts with a combo for a key that is used on non-US keyboard, then typing a key resulting from the combo takes a priority over the shortcut. For example `Alt + [` on an Italian Mac should still produce `[` (https://github.com/jupyterlab/jupyterlab/issues/15744).

This is we could distinguish between combos which fire [`beforeinput`](https://developer.mozilla.org/en-US/docs/Web/API/Element/beforeinput_event) and those which do not. Here is a PoC:


```html
<input id="the-input" placeholder="Enter some text" />
<p id="values"></p>
```

```js
const input = document.getElementById("the-input");
const log = document.getElementById("values");

input.addEventListener("keydown", maybeShortcut);
document.addEventListener("beforeinput", onbeforeinput);

let lastOnBeforeInput = null;

function onbeforeinput(e) {
  lastOnBeforeInput = e;
}

function maybeShortcut(e) {
  setTimeout(() => {
    const delta = lastOnBeforeInput.timeStamp - e.timeStamp;
    const data = JSON.stringify({
      code: e.code,
      shiftKey: e.shiftKey,
      metaKey: e.metaKey,
      altKey: e.altKey,
      tdelta: delta
    });
    if (lastOnBeforeInput && e.target === lastOnBeforeInput.target && Math.abs(delta) < 10) {
      log.textContent += `\ninput: ${data}`;
    } else {
      log.textContent += `\nshortcut: ${data}`;
    }
  }, 0)
}
```

it unfortunately relies on correct timestamps and these _might_ be obfuscated in Firefox, especially in some privacy modes - this needs checking.

**Comment by JasonWeill** [comment]
As discussed at the JupyterLab meeting today, we could potentially pop up a notification on first launch if we detect that the user's browser locale (which we should be able to detect for all modern browsers) is not en_US. We could direct the user to the settings editor. Having palettes of known-good keyboard shortcuts for particular keyboards might be useful.
